evap/rewards/forms.py

```diff
@@ -1,14 +1,108 @@
 
-from evap.rewards.models import RewardPointRedemptionEvent
+from evap.evaluation.models import UserProfile
+from evap.rewards.models import RewardPointRedemption, RewardPointRedemptionEvent
+from evap.rewards.tools import reward_points_of_user
 
 
 class RewardPointRedemptionEventForm(forms.ModelForm):
     class Meta:
         model = RewardPointRedemptionEvent
-        fields = ("name", "date", "redeem_end_date")
+        fields = ("name", "date", "redeem_end_date", "step")
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.fields["date"].localize = True
         self.fields["redeem_end_date"].localize = True
+
+
+class RewardPointRedemptionForm(forms.Form):
+    event = forms.ModelChoiceField(queryset=RewardPointRedemptionEvent.objects.all(), widget=forms.HiddenInput())
+    points = forms.IntegerField(min_value=0, label="")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if not self.initial:
+            return
+
+        self.fields["points"].validators.append(MaxValueValidator(self.initial["total_points_available"]))
+        self.fields["points"].widget.attrs["max"] = self.initial["total_points_available"]
+
+        self.fields["points"].validators.append(StepValueValidator(self.initial["event"].step))
+        self.fields["points"].widget.attrs["step"] = self.initial["event"].step
+
+        if self.initial["event"].step > 1:
+            self.fields["points"].help_text = _("multiples of {}").format(self.initial["event"].step)
+
+    def clean_event(self):
+        event = self.cleaned_data["event"]
+        if event.redeem_end_date < date.today():
+            raise ValidationError(_("Sorry, the deadline for this event expired already."))
+        return event
+
+
+class BaseRewardPointRedemptionFormSet(forms.BaseFormSet):
+    def __init__(self, *args, user: UserProfile, **kwargs) -> None:
+        self.user = user
+        super().__init__(*args, **kwargs)
+        self.locked = False
+
+    def get_form_kwargs(self, index):
+        kwargs = super().get_form_kwargs(index)
+        if not self.initial:
+            return kwargs
+        kwargs["initial"] = self.initial[index]
+        kwargs["initial"]["total_points_available"] = reward_points_of_user(self.user)
+        return kwargs
+
+    @contextmanager
+    def lock(self):
+        with transaction.atomic():
+            # lock these rows to prevent race conditions
+            list(self.user.reward_point_grantings.select_for_update())
+            list(self.user.reward_point_redemptions.select_for_update())
+
+            self.locked = True
+            yield
+            self.locked = False
+
+    def clean(self):
+        assert self.locked
+
+        if any(self.errors):
+            return
+
+        total_points_available = reward_points_of_user(self.user)
+        total_points_redeemed = sum(form.cleaned_data["points"] for form in self.forms)
```

Comment:
To make sure the comment from the other thread isn't lost: I propose to add
```suggestion
        total_points_redeemed = sum(form.cleaned_data["points"] for form in self.forms)
        assert all(form.cleaned_data["points"] > 0 for form in self.forms)
```
here

Reply:
We allow points to be zero if the user doesn't want to redeem points for an event. For values below 0, we validate the IntegerField with min_value=0. So, this additional check seems to be superfluous in my opinion.

Reply:
> We allow points to be zero if the user doesn't want to redeem points for an event.

Fair, I'm proposing `form.cleaned_data["points"] >= 0` then

> For values below 0, we validate the IntegerField with min_value=0. So, this additional check seems to be superfluous in my opinion.

It is relevant for security that >= 0 is validated somewhere. The contract of this snippet is "must have been validated before". Adding an assert would be a run-time checked codification of this contract. I don't regard this as validation, we'd still get an internal server error if it ever doesn't hold. In security-relevant questions, I strongly think we should do these kinds of contract-checks.